discovery_engine_core/providers/src/models/content.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{Error, NewscatcherArticle};
+use chrono::NaiveDateTime;
+use serde::{Deserialize, Serialize};
+use std::ops::Deref;
+use url::Url;
+
+/// A helper type used to ensure that, within the [`GenericArticle`] struct,
+/// we never use a URL which does not have a domain, such as `file:///foo/bar`.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct UrlWithDomain(Url);
+
+impl Deref for UrlWithDomain {
+    type Target = Url;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl UrlWithDomain {
+    pub fn new(url: Url) -> Option<Self> {
+        if url.domain().is_none() {
+            None
+        } else {
+            Some(UrlWithDomain(url))
+        }
+    }
+
+    #[allow(clippy::missing_panics_doc)]
+    pub fn domain(&self) -> &str {
+        self.0.domain().unwrap(/* constructor makes sure we have a domain */)
+    }
+
+    pub fn inner(&self) -> Url {
+        self.0.clone()
+    }
+}
+
+impl TryFrom<&str> for UrlWithDomain {
+    type Error = ();
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        Ok(UrlWithDomain::new(Url::parse(value).unwrap()).unwrap())
+    }
+}
+
+/// Represents a news that is delivered by an external content API.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct GenericArticle {
+    pub title: String,
+    pub snippet: String,
+    pub url: UrlWithDomain,
+    pub date_published: NaiveDateTime,
+    pub country: String,
+    pub language: String,
```

Comment:
❔Can we be sure that other providers also provide these attributes? does yt for example has this?

Reply:
As far as I understand, we'll always be querying for a certain market and/or language. YT does provide some localisation information, though only for language and location. It's a good point, though, I'll revisit whether it makes sense to remove `country` from here, depending on how we use it at the moment.

Reply:
I think we can leave this in for now. I will add a short ticket to clarify whether we're currently using `country`, since it's being exposed to blue. If they aren't we can remove it. wdyt?